Newsletters: Add subscriber modal

edanzer opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Addresses: 2236-gh-Automattic/dotcom-forge
Replaces this PR: https://github.com/Automattic/jetpack/pull/30970

## Proposed changes:
<!--- Explain what functional changes your PR includes -->

**Important for any testers: We've now put this behind a feature flag using the jetpack_subscriptions_modal_enabled filter. The most important thing to review and/or test is that the feature flag is correctly set to false, and will prevent the modal or modal related features or settings from loading when set to false.**

* Add a subscribe popup modal to frontend that allows readers to subscribe. This works by adding a Subscribe Modal template part to the site. This template part should only be created once. This template part contains the Subscriptions Block, and we load the entire template part inside the popup modal. 
* Add an option to enable the modal, and add the associated toggle on settings screen.
* Add the same new option to the settings endpoint - this change to the settings endpoint is not used in this PR, but will be used to add a separate option in calypso later for WP.com users.

**Limited initial exposure.**
As currently written the subscribe modal in **this feature should have very limited initial exposure**. 
   - First, even to see the option to turn on this feature, a site must have newsletter as site_intent (only done on [WP.com](http://wp.com/)) or have the Lettre theme active, and must have a block theme enabled (which likely excludes the majority of self-hosted jetpack sites). This is all just to see the option. 
   - Second, for the feature to become active, users must find the setting and actively enable the feature. And the setting for this is pretty hard to find, for better or worse.
   - In short, most users won't see the setting, and only those who enable the setting will see the modal. We should not have a situation where lots of users suddenly start seeing subscribe modals.
   - **If you are reviewing this PR, may be put special attention on confirm that I'm right and have correctly implemented all the checks above**. As long as these are all true, we have a lot of leeway and breathing space with this feature. 

### Outstanding questions
- We need to update the copy on the modal. It currently suggests it is only for protected posts. This needs design input.
- This PR is already large, and I'd like to avoid also updating the success behavior/notification when someone subscribes. But that part of the experience is currently not fluid. We could leave that as is, add it here, or update that via a separate PR before deploying this one. 
- This isn't working on Simple sites yet - I hadn't realized that we hide jetpack settings on simple sites. So still discussing solutions to that. 
- I have not deployed code before to the core jetpack plugin, so I'll welcome input on any aspect of this PR from seasoned Jetpack contributors. 
- I have not added tests. If possible, I'd prefer to add/update tests for this in a follow, again just given how big the PR is already.

**Modal looks like this:**

<img width="1507" alt="Screenshot on 2023-06-20 at 11-16-42" src="https://github.com/Automattic/jetpack/assets/21228350/f9817310-62dd-40bf-87c1-dc3f923e10ec">

**New option to enable the modal looks like this:**

<img width="1490" alt="Markup 2023-06-15 at 12 21 43" src="https://github.com/Automattic/jetpack/assets/21228350/fa58f538-ba8f-43c7-8d09-456580a945a1">

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
See 2236-gh-Automattic/dotcom-forge

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
Possibly? We are allowing our users to collect email addresses from their readers.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**Test on Simple Sites**
1. **Setup: Test Site**. You will need a test site that either has newsletter as the site_intent, or has the Lettre theme active. You can create one by starting at https://wordpress.com/setup/newsletter/intro if needed.
2. **Setup: Backend diff**. You'll need checkout this diff in your sandbox. It includes some extra code needed to enable Jetpack subscription functionality on WordPress.com: D114167-code
3. **Setup: Apply Jetpack PR**. In your sandbox run `bin/jetpack-downloader test jetpack add/subscriber-modal` to load this brand in your sandbox. Be sure to keep the backend diff from step 2 in place.
4. **Setup: Sandbox**. Be sure to sandbox public-api and the domain of your test site. 
5. **Setup: Add Filter**. As a form of feature flag, we've disabled this entire feature using a filter. To enable, go into your sandbox mu-plugins, and add the following line to your 0-sandbox.php file: 
   - `add_filter( 'jetpack_subscriptions_modal_enabled', '__return_true', 20 );`
7. **Setup: Calypso Dev + turn on option.** You need to enable the modal via a setting in Calypso. We've merged the PR for this, but that is also behind a feature flag. Checkout the trunk of calypso, run yarn start, go to the your site's admin are using http://calypso.localhost:3000/, navigate to Settings > Reading, scroll down to the Newletter settings, and enable the toggle for the subscriber modal.
8. **Setup: Add single post.** The modal will only show on a single post. If your test site does not have one, create one. 
9. **Test**: Go to front end of your website, go to a single post, and scroll. Shortly after you stop scrolling, you should see the subscribe modal popup. Confirm the modal loads and looks like the screenshot above. 
10. **Test**: Confirm that once you close the modal, it no longer loads (note: it will load again if you reload the page). 
11. **Test**: Enter an email and confirm you are subscribed, and the the behavior of the success notice is sane. Note: this is already a large PR, and I would prefer not to also update the success notification behavior of the subscribe modal. 

**Test on Atomic Site**
1. **Setup: WoA site**. You will need a WoA site. This site will need to either have newsletter as site_intent, or have the Lettre theme active.
2. **Setup: Jetpack beta tester**. Install and activate [the Jetpack beta plugin](https://jetpack.com/download-jetpack-beta/) on your site. Go to Jetpack > Jetpack Beta > Jetpack and enable the add/subscriber-modal branch.
3. **Setup: Add Filter**. As a form of feature flag, we've put this feature behind a filter. For the feature to show, you will need to use add_filter with the 'jetpack_subscriptions_modal_enabled' to return true. 
4. **Test**: Login to your site and go to Jetpack settings > Discussion. Enable Subscriptions and enable the Subscribe modal. Confirm that this toggle works, and that the update value persists (ie, try refreshing the page). 
5. **Test**: Go to front end of your website, and scroll. Shortly after you stop scrolling, you should see the subscribe modal popup. Confirm the modal loads and looks like the screenshot above. 
6. **Test**: Confirm that once you close the modal, it no longer loads (note: it will load again if you reload the page). 
7. **Test**: Enter an email and confirm you are subscribed, and the the behavior of the success notice is sane. Note: this is already a large PR, and I would prefer not to also update the success notification behavior of the subscribe modal. 

**Test on self-hosted Jetpack site (ie, just use your existing jetpack dev environment)**
1. **Setup: Normal Jetpack dev environment**. Use your normal Jetpack dev environment. Checkout this branch and run any needed build commands to build jetpack locally.
2. **Setup: Jurassic Tube.** This functionality requires a connected site, so you'll need to set up a jurassic tube site and 
3. **Setup: Lettre theme.** Activate the Lettre theme on your dev site. 
4. **Setup: Add Filter**. As a form of feature flag, we've put this feature behind a filter. For the feature to show, you will need to use add_filter with the 'jetpack_subscriptions_modal_enabled' to return true. 
5. **Test**: Login to your site and go to Jetpack settings > Discussion. Enable Subscriptions and enable the Subscribe modal. Confirm that this toggle works, and that the update value persists (ie, try refreshing the page). 
6. **Test**: Go to front end of your website, and scroll. Shortly after you stop scrolling, you should see the subscribe modal popup. Confirm the modal loads and looks like the screenshot above. 
7. **Test**: Confirm that once you close the modal, it no longer loads (note: it will load again if you reload the page). 
8. **Test**: Enter an email and confirm you are subscribed, and the the behavior of the success notice is sane. Note: this is already a large PR, and I would prefer not to also update the success notification behavior of the subscribe modal.

edanzer committed:
[not verified] Add subscriber modal

edanzer committed:
[not verified] changelog

edanzer committed:
[not verified] Add option to enable subscribe modal

edanzer committed:
[not verified] changelog

edanzer committed:
[not verified] Remove test change for now

edanzer committed:
[not verified] Remove old option from sync list

edanzer commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+<?php
```
I should probably namespace and/or move code in this file into a class. At the very least, I will remove the wpcom prefixes.

simison replied:
Jetpack prefixes are prolly enough? I'm not sure how fond we're using namespaces in Jetpack. @jeherve ?

jeherve replied:
A namespace would be okay. We use them in a lot of places in the codebase.

edanzer replied:
Following what seems to be convention in the module folders, I did not add a namespace, but did update the file to be a class, so all method names are now encapsulated within that class.

crisbusquets commented:
Hi @edanzer !

Great work, and thanks for sharing the video in Slack too :)

**Modal behavior**
Would it be possible that the modal only appears in posts? Now that I see it, I find it odd that it's on the homepage, where there are already two subscriber blocks. 

**Copy**
The copy will be ok if we can change this behavior. If it's not possible, I'd change the copy to:

Heading
`Discover more from {SiteName}`

Text
`Subscribe to the newsletter to keep reading and get access to the full archive.`

Link to close (in any copywriting option we use)
`Continue reading`

[Figma file](h8tMpJlCqNFemEerU9owHI-fi-522_9013) and screenshot: 

<img width="572" alt="Screenshot 2023-06-16 at 09 31 42" src="https://github.com/Automattic/jetpack/assets/25607244/4a865dc9-e738-4479-99ec-522da814803f">


**Success screen**
I'd use this, it's not ideal but it might work for now: https://github.com/Automattic/jetpack/issues/30578

[Figma file](h8tMpJlCqNFemEerU9owHI-fi-522_9013) and screenshot:

<img width="570" alt="Screenshot 2023-06-16 at 09 31 26" src="https://github.com/Automattic/jetpack/assets/25607244/5d1adbcc-2e4b-40fa-ad98-e06470a4bae5">


**Questions**
- Is it possible to close de modal when clicking outside it?
- The modal uses the typography on the design, not the one that the user has in the theme, right?

🚀

simison commented:
> The modal uses the typography on the design, not the one that the user has in the theme, right?

That's an interesting one; the subscribe block does inherit theme styles, why not do the same with paragraph and heading?

simison commented on projects/plugins/jetpack/_inc/client/discussion/subscriptions.jsx:
```diff
@@ -134,6 +139,17 @@ class SubscriptionsComponent extends React.Component {
+								}
+								toggling={ this.props.isSavingAnyOption( [ 'sm_enabled' ] ) }
+								onChange={ this.handleSubscribeModalToggleChange }
+								label={ __( 'Enable the popup Subscribe Modal', 'jetpack' ) }
```
Copy feels a bit clunky, but I don't have alternatives to suggest now. Non blocking merge.

jeherve replied:
I would agree. The capitalization was what jumped at me. Maybe this could at least be:

```suggestion
								label={ __( 'Enable the popup subscription modal', 'jetpack' ) }
```

lezama replied:
Erick is probably following the pattern of the two previous settings.

<img width="555" alt="Screenshot 2023-07-05 at 18 12 52" src="https://github.com/Automattic/jetpack/assets/104869/8f8168a8-68b2-44ea-8617-88fc4352d7a1">

simison commented on projects/plugins/jetpack/_inc/client/discussion/subscriptions.jsx:
```diff
@@ -134,6 +139,17 @@ class SubscriptionsComponent extends React.Component {
+								toggling={ this.props.isSavingAnyOption( [ 'sm_enabled' ] ) }
+								onChange={ this.handleSubscribeModalToggleChange }
+								label={ __( 'Enable the popup Subscribe Modal', 'jetpack' ) }
+							/>
```
Can we have a link to edit the modal in the focused template part editor somewhere near this toggle? Ok in follow-up PR as well.

edanzer replied:
Yup, I will handle this in a follow up PR.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.css:
```diff
@@ -0,0 +1,45 @@
+#wpcom-subscribe-modal {
```
Nit; can be `jetpack-` prefix but no biggie if wpcom.

jeherve replied:
I would agree with the prefix. I would also suggest, if possible, that we use a class for all those changes if possible, so site owners have it a bit easier if they want to customize the look of the pop-up, by just targeting the (more specific) ID.

edanzer replied:
I've updated this from id to class, and from `wpcom-` to `jetpack-`.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.css:
```diff
@@ -0,0 +1,45 @@
+	position: relative;
+	visibility: hidden;
+    top: 100%;
+	background-color: #fefefe;
```
Some mixed tabs/spaces white spacing here and elsewhere in the file.

edanzer replied:
Fixed!

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.js:
```diff
@@ -0,0 +1,27 @@
+window.onload = function () {
+	const modal = document.getElementById( 'wpcom-subscribe-modal' );
```
Let's bail out early if the element isn't found for any reason.

edanzer replied:
Added this.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.js:
```diff
@@ -0,0 +1,27 @@
+window.onload = function () {
```
There's a GB package for this: https://github.com/WordPress/gutenberg/tree/trunk/packages/dom-ready#readme

It has a bit more complete version of onload:

https://github.com/WordPress/gutenberg/blob/trunk/packages/dom-ready/src/index.js#L31-L45

edanzer replied:
Added this.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.css:
```diff
@@ -0,0 +1,45 @@
+	text-align: center;
+}
+
+#wpcom-subscribe-modal #close a {
```
We might need to get more defensive with styles here in case we want theme styles not leak in? Worth testing with dozen different themes! [See plugin](https://wordpress.org/plugins/back-to-the-theme/)

edanzer replied:
Handy tool! I did try this. The first thing I found the modal wasn't loading at all on most themes due to a JS error with the wp object not being available. So I pushed a fix for that - just requiring wp-components when enqueuing.

I then used the tool, and below is the result. Obviously the style varies quite a bit. This is intentional for now, and reflects how WordPress is supposed to work in the block editor / FSE era. In short, the modal picks up default theme styling (fonts, colors, etc). We'd need to decide if we want to override those. My sense is probably not. 

The one thing that does need fixing is there are several cases where the font is dark on the button, and thus not visible. Given buttons are generally darker colors, it could make sense to force a white font on the button? 

![screencapture-erick-jurassic-tube-wp-admin-tools-php-2023-06-27-11_15_10](https://github.com/Automattic/jetpack/assets/21228350/caccdd37-3407-477e-9656-e3db187b5721)

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+ * Subscribe modal popup feature
+ *
+ * @package automattic/jetpack-mu-wpcom
+ * @since 2.2.1
```
Package name needs updating/removing

Needs some `$` variable here that gets automatically replaced with right version later I think.

jeherve replied:
+1. That would give us

```suggestion
 * @since $$next-version$$
```

For more information:
https://github.com/Automattic/jetpack/blob/2aac5857f66d07bfda1e08da7db299b9aaf63db9/projects/packages/README.md#package-version-annotations

edanzer replied:
I've updated this!

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+						echo wp_kses( render_block( $block ), $allowed_html );
+					}
+					?>
+					<div id="close"><a href="#">Close</a></div>
```
"Close" needs wrapping in translation function.

edanzer replied:
FYI - text for this has been moved inside the subscribe modal template part. So any translation will need to be handled there.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @return bool
+ */
+function should_enable_subscriber_modal() {
```
We can also limit to WP.com on first release?

Rest of the limits won't make sense for Jetpack sites, but for .com they could be reasonable start. And then release a bit wider next up?

jeherve replied:
Maybe we could add a filter in here, so folks wanting to test can use the filter to toggle on the feature?

edanzer replied:
@jeherve - interesting thought. Can you clarify or maybe link to an example of how that might work in jetpack for a situation like this? Not sure if we'd do this via, say, constant in wp-config or an actual WP filter or some other mechanism?

jeherve replied:
Basically something like this:

```diff
diff --git a/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php b/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php
index 332abe2165..b77837781c 100644
--- a/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php
+++ b/projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php
@@ -187,7 +187,16 @@ class Jetpack_Subscribe_Modal {
 		$is_modal_enabled   = get_option( 'sm_enabled' ) || false;
 		$is_fse_theme       = wp_is_block_theme();
 
-		return ( $is_lettre || $is_newsletter_site ) && $is_fse_theme && $is_modal_enabled;
+		$should_enable_subscriber_modal = ( $is_lettre || $is_newsletter_site ) && $is_fse_theme && $is_modal_enabled;
+
+		/**
+		 * Allows force-enabling or disabling the Subscriptions modal.
+		 *
+		 * @since $$next-version$$
+		 *
+		 * @param bool $should_enable_subscriber_modal Whether the Subscriptions modal should be enabled.
+		 */
+		return (bool) apply_filters( 'jetpack_subscriptions_modal_enabled', $should_enable_subscriber_modal );
 	}

```

You would then be able to force-enable the modal by adding the following in a functionality plugin:

```php
add_filter( 'jetpack_subscriptions_modal_enabled', '__return_true' );
```

edanzer replied:
@jeherve - Thanks for the details. I've added this. I find myself wondering if we should also add a filter for users to add their own modal content, vs just using the new subscribe modal template part.  I'd do that later as a separate PR.

As a side note, a lot of my recent dev work has been on WordPress.com, and in that environment we think a lot less about leveraging these WP apis to make code behavior customizable by other devs. So this was a great general lesson to be thinking much more about that when working on Jetpack or other code that will be delivered via WordPress.org.

simison replied:
I think it's fine if they can modify the modal via template editor and no need to modify via code.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+ * @return string
+ */
+function get_subscribe_template_title() {
+	return 'Subscribe Modal';
```
Translation function

edanzer replied:
Added.

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+function get_subscribe_template_content() {
+	return '<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|40","right":"var:preset|spacing|40","bottom":"var:preset|spacing|40","left":"var:preset|spacing|40"}}},"layout":{"type":"constrained"}} -->
+	<div class="wp-block-group" style="padding-top:var(--wp--preset--spacing--40);padding-right:var(--wp--preset--spacing--40);padding-bottom:var(--wp--preset--spacing--40);padding-left:var(--wp--preset--spacing--40)"><!-- wp:heading {"textAlign":"center"} -->
+	<h2 class="wp-block-heading has-text-align-center">This post is for subscribers</h2>
```
Translation function.

edanzer replied:
Is there any weirdness about trying to use a translation function inside of block markup?

edanzer replied:
Note: I've actually moved the text/markup for the Close button inside the template part now too, so same question will apply on that.

edanzer replied:
Updated see https://github.com/Automattic/jetpack/pull/31393#discussion_r1244160182

simison commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @return void
+ */
+function wpcom_create_subscribe_template() {
```
Do we need any function checks and bail out, or everything we need is supported in WP versions that Jetpack supports? Assume Gutenberg plugin gets disabled.

edanzer replied:
Good question. I have added a check for wp_is_block_theme() based on separate feedback below. As is, we're now checking if it's a block theme, if the modal option is turned on, and if (the site intent is newsletter or the theme is lettre). Off the top, I'd think that would be enough checks, though I'm happy to add others if we think doing so is warranted.

simison commented:
> Success screen
> I'd use this, it's not ideal but it might work for now: https://github.com/Automattic/jetpack/issues/30578

This success screen update is a good separate generic update to the block as well; and I'd also consider separate PR for it.

jeherve commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -461,6 +476,29 @@ public function subscription_comment_subscribe_setting() {
+		<p class="description">
+			<input type="checkbox" name="sm_enabled" id="jetpack-subscribe-modal" value="1" <?php checked( $sm_enabled, 1 ); ?> />
+			<?php
+			echo wp_kses(
+				__(
+					'Show a popup subscribe modal to readers.',
+					'jetpack'
+				)
+			);
```
I don't know that we need `wp_kses` here. It could just be 

```suggestion
			esc_html_e( 'Show a popup subscribe modal to readers.', 'jetpack' );
```

edanzer replied:
Yes, this is correct. A bit of copy/past going on here. Fixed.

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+			)
+		);
+
+		$subscribe_template = reset(
+			array_filter(
+				$posts,
+				function ( $post ) {
+					return $post->post_name === get_subscribe_template_slug();
+				}
+			)
+		);
+
+		$blocks = parse_blocks( $subscribe_template->post_content );
```
Should we bail here on sites that won't have / support template parts?

edanzer replied:
Good thought. I believe all FSE themes should support template parts? Assuming yes, I've added a check with `wp_is_block_theme()` here: 

https://github.com/Automattic/jetpack/pull/31393/files#diff-bcd4c897206a1f66102ef44f0e272985d67be37c91dfacbbd6458d46d66ac72eR188

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/subscribe-modal.php:
```diff
@@ -0,0 +1,196 @@
+ */
+function wpcom_enqueue_subscribe_modal_assets() {
+	if ( should_enable_subscriber_modal() && ! is_admin() ) {
+		wp_enqueue_style( 'subscribe-modal-css', plugins_url( 'subscribe-modal.css', __FILE__ ), array(), JETPACK__VERSION );
```
You could add this new stylesheet to `concatenated_style_handles` in `class.jetpack.php` so the contents of the file can be added to the concatenated `jetpack.css` that's loaded for all connected Jetpack sites.

edanzer replied:
I've added this. To confirm, I still enqueue the new stylesheet like normal too? Based on what I'm seeing, that seems to be case.

jeherve replied:
Yes, you continue to enqueue the stylesheet as usual. We take care of dequeuing it when the concatenated version has to be enqueued instead.

edanzer committed:
[not verified] Update subscribe modal file to class

edanzer committed:
[not verified] Update markup from wpcom- id to jetpack- class

edanzer committed:
[not verified] Fix mixed spaces/tabs in css file

edanzer committed:
[not verified] Skip modal js if modal element not found

edanzer committed:
[not verified] Update to use domReady()

edanzer committed:
[not verified] Make php @since dynamic

edanzer committed:
[not verified] Add translate functions

edanzer committed:
[not verified] Replace unneded wp_kses method

edanzer committed:
[not verified] Add css to jetpack concatenated css

edanzer committed:
[not verified] Add check for fse themes

edanzer committed:
[not verified] Update modal content and styling

edanzer commented:
@crisbusquets 

I've addressed most of your feedback, and updated the screenshot in the PR description for how the modal looks now. 

> Would it be possible that the modal only appears in posts?

This has been done!

> The copy will be ok if we can change this behavior. If it's not possible, I'd change the copy to:

Even though we only show on posts, I've gone ahead and updated the copy to what you suggest above anyways. The way the modal works, we don't (for now) prevent users from reading - ie, it is not limited to subscribers. The new copy you suggested above reads nicely and seems a bit more versatile. 

> Success screen/behavior

If it's ok, I'm going to try to tackle this in a separate PR that updates the subscribe block itself. 

> Is it possible to close de modal when clicking outside it?

Yes, the modal will close if you click anywhere outside of it!

> The modal uses the typography on the design, not the one that the user has in the theme, right?

It is actually using the theme fonts right now. This includes not just font family, but also certain things like font weight, which have limited options in the block editor. I'm open to changing that. Two notes. First, it may makes sense to keep fonts consistent for each site - ie, a site that all all blocky serif fonts may want those in the modal. Second, we have limited ability to add custom fonts within the core blocks in the block editor. It's possible we'd need to override these via separate CSS, which just means users could not edit the font family short of writing their own custom css. 

> Figma

For some reason, the figma file links you provided don't work. Something's not quite working with the dynamic links. I'm trying to keep accurate, but I'm not quite sure how to find the figma files without the direct links above.

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,236 @@
+		return '<!-- wp:group {"style":{"spacing":{"padding":{"top":"50px","bottom":"50px","left":"20px","right":"20px"}}},"layout":{"type":"constrained"}} -->
+		<div class="wp-block-group" style="padding-top:50px;padding-right:20px;padding-bottom:50px;padding-left:20px"><!-- wp:group {"style":{"dimensions":{"minHeight":"0px"},"spacing":{"blockGap":"8px"}},"layout":{"type":"flex","flexWrap":"wrap","justifyContent":"center"}} -->
+		<div class="wp-block-group" style="min-height:0px"><!-- wp:heading {"style":{"typography":{"fontStyle":"normal","fontWeight":"600","fontSize":"26px"},"layout":{"selfStretch":"fit","flexSize":null}}} -->
+		<h2 class="wp-block-heading" style="font-size:26px;font-style:normal;font-weight:600">Discover more from</h2>
```
should we make these translatable?

edanzer replied:
Yes, I'd posted a question above. Think there's any weirdness in trying to translate content that is ultimately inserted as post content in the block editor?

edanzer replied:
Since we weren't sure, I just tested this. It works. Having the translation methods as part of the markup here doesn't negatively affect creating the template. So I've pushed the change adding translations.

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,236 @@
+	 * Jetpack_Subscribe_Modal class constructor.
+	 */
+	public function __construct() {
+		add_action( 'wp_loaded', array( $this, 'wpcom_create_subscribe_template' ) );
```
should we move the `if $this->should_enable_subscriber_modal()` check up here?

edanzer replied:
That makes sense to me. Might as well do the check early as possible and run as little unneeded code as possible. I'll update that.

edanzer replied:
I updated this. I looked at moving the other checks for loading the frontend modal markup and JS, but the nature of those checks is that they need to happen later in the WordPress loading process.

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,236 @@
+			$this->should_enable_subscriber_modal() &&
+			! is_admin() &&
+			is_single() &&
+			'post' == get_post_type()
```
```suggestion
			'post' === get_post_type()
```

edanzer replied:
Good catch, will update.

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,236 @@
+				array(
+					'post_type'   => 'wp_template_part',
+					'post_status' => 'publish',
+					'numberposts' => -1,
```
I think you can pass 

```php
'name' => $this->get_subscribe_template_slug(),
``` 

so you don't need to do the `reset` call bellow

edanzer replied:
Good call @lezama! Updated this. While I was at it, I also added a check to confirm we found posts in this query before we proceed to try adding the template modal.

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,236 @@
+				)
+			);
+
+			$blocks = parse_blocks( $subscribe_template->post_content );
```
I think if we do `apply_filters('the_content', $post->post_content);` we can avoid doing the loop and `wp_kses`

edanzer replied:
Hey @lezama - Couple questions on this one...

First...  Basically, we're taking some content from an editor template, content which is formatted in block format, and outputting outside the normal context, in a custom modal. My understanding is that you'd still need to correctly parse/convert the block markup to regular html mark up. Would the method you're describing somehow do that as well?

Second... I may be rusty on my filters, but apply_filter() I think just adds a new filter. Not sure how that would resolve the main issue just above. Did you mean to suggest add_filter()? Even in that case, I wouldn't think add_filter( 'the_content', $my_content ) wouldn't work here either. That filters the post content, but doesn't solve the issue of converting block markup to html markup.

lezama committed:
[not verified] Update projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php

edanzer committed:
[not verified] Simplify conditionals

edanzer committed:
[not verified] Ensure wp object is loaded on frontend

edanzer committed:
[not verified] Add filter for should_enable_subscriber_modal()

edanzer committed:
[not verified] Add translations to subscribe modal content

edanzer committed:
[not verified] Simply post filtering for subsribe modal template

edanzer committed:
[not verified] Add check for empty posts before adding modal

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,251 @@
+				array(
+					'post_type'   => 'wp_template_part',
+					'post_status' => 'publish',
+					'numberposts' => -1,
```
why not  `'numberposts'   => 1,`?

edanzer replied:
While it shouldn't happen, I had an instance today where this query returned more than one post, and the right post was the second / more recent one. Somehow, and old version was still there matching the query. That's why the combo of -1 here (which is any matches) and end() just below, which grabs the most recent match.

lezama replied:
oh, we could add `order` to the query then.

edanzer commented:
FYI - I'm looking into the failing PHP unit tests now. Looks like those are failing because add a whitelist option but didn't updated tests to include that.

edanzer committed:
[not verified] Ensure subscribe post request goes through

edanzer committed:
[not verified] Fix unit tests for sync settings

edanzer committed:
[not verified] Fix JS tests

edanzer committed:
[not verified] Fix version tests

lezama commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -183,7 +183,16 @@ public function should_enable_subscriber_modal() {
+		*
+		* @param bool $should_enable_subscriber_modal Whether the Subscriptions modal should be enabled.
+		*/
+		return (bool) apply_filters( 'jetpack_subscriptions_modal_enabled', $should_enable_subscriber_modal );
```
perfect

lezama committed:
[not verified] add feature flag

edanzer commented:
I just tested the new 'jetpack_subscriptions_modal_enabled' that we've added as a feature flag. I also updated testing instructions for others to set that now.

edanzer committed:
[not verified] Add test for modal conditions

lezama committed:
[not verified] fix failing test

jeherve commented on projects/plugins/jetpack/_inc/client/discussion/subscriptions.jsx:
```diff
@@ -134,6 +139,17 @@ class SubscriptionsComponent extends React.Component {
 									'jetpack'
 								) }
 							/>
+							<ToggleControl
```
I think it may be best to hide this with the feature flag. Otherwise folks will toggle that option, expect something to happen, and reach out to support when nothing happens.

You can check #31479 for an example of a PR that adds the necessary tooling to pull the feature flag from PHP into the dashboard here.

lezama replied:
[Done!](https://github.com/Automattic/jetpack/pull/31393/commits/e5134959dabfce6e9e36bb6893b8e1c4c1feac24)

jeherve commented on projects/packages/sync/src/class-package-version.php:
```diff
@@ -12,7 +12,7 @@
 class Package_Version {
 
-	const PACKAGE_VERSION = '1.50.2-alpha';
+	const PACKAGE_VERSION = '1.51.1-alpha';
```
Looks like you'll need to update this PR with latest trunk once more, and run the script to update the versions.

jeherve commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -348,6 +348,21 @@ public function configure() {
 
+		/** Enable Subscribe Modal */
+
+		add_settings_field(
```
I think this should also be hidden behind the feature flag.

lezama replied:
[done](https://github.com/Automattic/jetpack/commit/41c5af8c58e3698130b7906748bdacffcb8ccbdb), where these show up?

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+							echo wp_kses( render_block( $block ), $allowed_html );
+						}
+						?>
+						<Dashicon icon="minus" className="jp-idc__idc-screen__card-action-separator" />
```
This isn't valid HTML, so it won't be displayed properly I'm afraid.

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+	 *
+	 * @return void
+	 */
+	public function wpcom_create_subscribe_template() {
```
I wonder if this could be triggered when the toggle is enabled in the Jetpack Settings. This may prove better for performance, since later on we wouldn't have to check for an existing post on each `wp_loaded`. What do you think?

simison replied:
It might still be important to do the check outside the toggle; thinking situations where we add the option by default onto the site when site is created.

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+	public function wpcom_enqueue_subscribe_modal_assets() {
+		if ( $this->is_front_end_single_post() ) {
+			wp_enqueue_style( 'subscribe-modal-css', plugins_url( 'subscribe-modal.css', __FILE__ ), array(), JETPACK__VERSION );
+			wp_enqueue_script( 'subscribe-modal-js', plugins_url( 'subscribe-modal.js', __FILE__ ), array( 'wp-components' ), JETPACK__VERSION, true );
```
Do we need `wp-components` to be loaded here? Could we do without it, to avoid loading too many additional things on the frontend?

simison replied:
This would be important to solve — can be a follow-up PR.

simison replied:
Actually we shouldn't need it, looking at `subscribe-modal.js`, but we do need `domReady` which is a tiny script.

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+	 * @return bool
+	 */
+	public function should_enable_subscriber_modal() {
+		$is_lettre          = get_option( 'stylesheet' ) === 'lettre';
+		$is_newsletter_site = get_option( 'site_intent' ) === 'newsletter';
+		$is_modal_enabled   = get_option( 'sm_enabled' ) || false;
```
Instead of `|| false` here, you should be able to provide a default option in your `get_option` call:

```suggestion
		$is_modal_enabled   = get_option( 'sm_enabled', false );
```

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+	 * @return bool
+	 */
+	public function is_front_end_single_post() {
+		return ! is_admin() && is_single() && 'post' === get_post_type();
```
You should be able to use `is_singular` here:

```suggestion
		return ! is_admin() && is_singular( 'post' );
```

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+		return '<!-- wp:group {"style":{"spacing":{"padding":{"top":"50px","bottom":"50px","left":"20px","right":"20px"}}},"layout":{"type":"constrained"}} -->
+		<div class="wp-block-group" style="padding-top:50px;padding-right:20px;padding-bottom:50px;padding-left:20px"><!-- wp:group {"style":{"dimensions":{"minHeight":"0px"},"spacing":{"blockGap":"8px"}},"layout":{"type":"flex","flexWrap":"wrap","justifyContent":"center"}} -->
+		<div class="wp-block-group" style="min-height:0px"><!-- wp:heading {"style":{"typography":{"fontStyle":"normal","fontWeight":"600","fontSize":"26px"},"layout":{"selfStretch":"fit","flexSize":null}}} -->
+		<h2 class="wp-block-heading" style="font-size:26px;font-style:normal;font-weight:600">' . __( 'Disover more from', 'jetpack' ) . '</h2>
```
```suggestion
		<h2 class="wp-block-heading" style="font-size:26px;font-style:normal;font-weight:600">' . __( 'Discover more from', 'jetpack' ) . '</h2>
```

jeherve replied:
I should add that this may prove difficult to translate for translators, since the site title is outside of the string. In languages where the site title would come before "from", it will be a problem.
It would be nice to have the site title part of that string instead.

jeherve commented on projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php:
```diff
@@ -0,0 +1,261 @@
+		<!-- /wp:paragraph -->
+		
+		<!-- wp:paragraph {"align":"center","style":{"typography":{"fontSize":"15px"},"spacing":{"margin":{"bottom":"20px","top":"0px"}}}} -->
+		<p class="has-text-align-center" style="margin-top:0px;margin-bottom:20px;font-size:15px">and get access to the full archive.</p>
```
We'll probably want to make translatable.

simison replied:
Same note about ease of translating https://github.com/Automattic/jetpack/pull/31393#discussion_r1253363857

jeherve reviewed (dismissed):
Nice work so far, thanks for sticking with it in this massive PR :)

I gave it a quick try and left some notes below.

*********

I'm seeing a number of notices when the modal is displayed:

```
PHP Notice:  Undefined index:  in /wp-includes/blocks/template-part.php on line 238
PHP Warning:  Cannot modify header information - headers already sent by (output started at /wp-includes/blocks/template-part.php:238) in /wp-includes/pluggable.php on line 1427
```

lezama committed:
[not verified] Update projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

lezama committed:
[not verified] Update projects/plugins/jetpack/modules/subscriptions/subscribe-modal/class-jetpack-subscribe-modal.php

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

lezama commented:
Thanks @jeherve! I will be working on your feedback since Erick is AFK this week.

> I'm seeing a number of notices when the modal is displayed:

I am not getting these, anything I should do in particular to reproduce?

simison committed:
[not verified] Use lower case "subscription modal" copy

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

simison commented:
In my standalone Jetpack with Gutenberg disabled I'm getting this visually, but I still need to try see if something didn't just build correctly.

<img width="669" alt="image" src="https://github.com/Automattic/jetpack/assets/87168/d73d7b32-2243-4848-80f8-0e53c02d2088">

jeherve commented:
> I am not getting these, anything I should do in particular to reproduce?

I got them on a fresh JN site with current WP and no Gutenberg. I didn't do anything special I think.

lezama commented:
@scruffian what do you think of [this way of adding a template part](https://github.com/Automattic/jetpack/pull/31393/files#diff-bcd4c897206a1f66102ef44f0e272985d67be37c91dfacbbd6458d46d66ac72eR52). Anything we could do differently?

Also any ideas on why the template part wouldn't show up on my test site full site editor 😅 ?

lezama committed:
[not verified] update should_enable_subscriber_modal

lezama committed:
[not verified] typo

lezama committed:
[not verified] feature flag modal settings

lezama committed:
[not verified] Add feature flag to core settings

lezama committed:
[not verified] fix typo

scruffian commented:
> @scruffian what do you think of [this way of adding a template part](https://github.com/Automattic/jetpack/pull/31393/files#diff-bcd4c897206a1f66102ef44f0e272985d67be37c91dfacbbd6458d46d66ac72eR52). Anything we could do differently?

I think you should look into doing this via the API - there's an enpoint at `/wp-json/wp/v2/templates?_locale=user`, which you should be able to post to, to create a new template. Let me know if you need more help with that :)

scruffian commented:
Another option is adding templates via a filter, like I prototype here: https://github.com/WordPress/create-block-theme/pull/124

This avoids creating stuff in a user's database unless they really need it.

lezama committed:
[not verified] add template with filters

lezama committed:
[not verified] Some more improvements

lezama committed:
[not verified] reorganize code

lezama committed:
[not verified] use HereDoc

lezama committed:
[not verified] fix indentation

lezama committed:
revert unrelated changes

lezama committed:
remove no more relevant changelog

lezama committed:
delete inaccurate changelog

lezama commented:
After @simison and @scruffian feedback I cleaned up the code inspired by [Ben's suggestion](https://github.com/Automattic/jetpack/pull/31393#issuecomment-1625386473). I will follow up with a PR to make the template part show up in the site editor.

@jeherve I think I went through all your feedback, except one thing, the mess with the composer.lock files and package versions. I would love your help with that 🙏

lezama committed:
Make the template part work

lezama committed:
Merge branch 'add/subscriber-modal' of github.com:Automattic/jetpack into add/subscriber-modal

lezama committed:
something like this?

lezama committed:
would this fix the tests?

lezama committed:
oops

lezama committed:
Show settings when  'sm_enabled' == false

lezama committed:
probably not correct and overengineering

edanzer committed:
Update filter for testing

edanzer committed:
Update jetpack_subscriptions_modal_enabled filter docs

edanzer committed:
Update feature flag comment for readability

oskosk committed:
Merge branch 'trunk' into add/subscriber-modal

oskosk commented:
The changelogger check is failing. When this happens, a rebase against trunk (`git merge trunk`) and the command shown in the files view for this PR should be run, followed by a commit, and a new push. I'm doing it for you now.

<img width="985" alt="image" src="https://github.com/Automattic/jetpack/assets/746152/df62eecb-20eb-4f57-ac68-13cdee75b2fd">

oskosk committed:
versions update

oskosk commented:
Pushed rebase in 67f117c3f2832008cc87837e5212deda67a49202 and versions update in 8fe478c1a2c64e2b6d7c1fc9af58b1700dcba41a

edanzer committed:
Update tests to reflect current filter logic

edanzer commented:
The new tests written for this PR were failing due to latest changes. I've updated them to reflect the current logic for the feature flag and for should_load_subscriber_modal().

Valecia2023 commented:
circular-form.jpg-photo-message

simison committed:
Merge branch 'trunk' into add/subscriber-modal

simison committed:
Update doc about filter to be "true" not "false"

simison committed:
Match toggle copy with what we have in Calypso

simison commented on projects/plugins/jetpack/_inc/client/discussion/subscriptions.jsx:
```diff
@@ -134,6 +141,27 @@ class SubscriptionsComponent extends React.Component {
+										onChange={ this.handleSubscribeModalToggleChange }
+										label={ __( 'Enable subscriber modal', 'jetpack' ) }
+									/>
+									<p className="jp-form-setting-explanation">
```
I used separate `p` here because `help={}` prop just doesn't look great at all for a longer text. I copied the pattern from elsewhere in Jetpack settings:

**help prop**
<img width="1091" alt="Screenshot 2023-07-13 at 11 38 34" src="https://github.com/Automattic/jetpack/assets/87168/062405a8-0727-447f-ac52-164af82c587f">

**paragraph**
<img width="1078" alt="Screenshot 2023-07-13 at 11 44 01" src="https://github.com/Automattic/jetpack/assets/87168/43939d26-dafd-41a9-9311-5a90566af154">

simison reviewed (dismissed):
This is in good shape now! I've tested a bunch. Since this is behind feature flags, we can safely merge and continue polishing (and releasing soon!) in follow-up PRs.

<img width="1078" alt="Screenshot 2023-07-13 at 11 44 01" src="https://github.com/Automattic/jetpack/assets/87168/5a58bb5c-d6f5-493f-98db-d49c457c1ae1">

<img width="1057" alt="image" src="https://github.com/Automattic/jetpack/assets/87168/f8ed46d8-a113-4d7c-89e4-d5b54ef8eeec">

In follow-up PRs we should:
- Add "edit modal" link next to the toggle
- Tweak logic when the modal shows up; site owners, already subscribed, if the post is paid/subscribe walled already in which case they see the same content inline already, etc.

lezama committed:
fix package version

jeherve commented on projects/plugins/jetpack/modules/subscriptions.php:
```diff
@@ -461,6 +479,22 @@ public function subscription_comment_subscribe_setting() {
+	 */
+	public function subscribe_modal_setting() {
+
+		$sm_enabled = get_option( 'sm_enabled', 1 );
+		?>
+
+		<p class="description">
+			<input type="checkbox" name="sm_enabled" id="jetpack-subscribe-modal" value="1" <?php checked( $sm_enabled, 1 ); ?> />
```
Coming back to this today; shouldn't the default / fallback here be `0` everywhere, since we do not enable the modal by default?
